File: visualizations/quantum-hall-band-topology/static/physics.mjs
const TAU = 2 * Math.PI;
const EPSILON = 1e-11;

function assertFinite(value, name) {
  if (!Number.isFinite(value)) throw new RangeError(`${name} must be finite`);
}

function norm3(vector) {
  return Math.hypot(vector.x, vector.y, vector.z);
}

function dot3(a, b) {
  return a.x * b.x + a.y * b.y + a.z * b.z;
}

function cross3(a, b) {
  return {
    x: a.y * b.z - a.z * b.y,
    y: a.z * b.x - a.x * b.z,
    z: a.x * b.y - a.y * b.x,
  };
}

/** Coefficient vector for H(k) = d(k).sigma in the declared QWZ convention. */
export function qwzD(kx, ky, parameters = {}) {
  const mass = parameters.mass ?? -1;
  const hopping = parameters.hopping ?? 1;
  const trBreaking = parameters.trBreaking ?? 1;
  for (const [value, name] of [
    [kx, "kx"],
    [ky, "ky"],
    [mass, "mass"],
    [hopping, "hopping"],
    [trBreaking, "trBreaking"],
  ]) assertFinite(value, name);

  return {
    x: hopping * Math.sin(kx),
    y: -trBreaking * Math.sin(ky),
    z: mass + Math.cos(kx) + Math.cos(ky),
  };
}

export function qwzEnergy(kx, ky, parameters = {}) {
  return norm3(qwzD(kx, ky, parameters));
}

export function unitD(kx, ky, parameters = {}) {
  const vector = qwzD(kx, ky, parameters);
  const length = norm3(vector);
  if (length < EPSILON) return null;
  return {x: vector.x / length, y: vector.y / length, z: vector.z / length};
}

/** Berry curvature of the occupied (negative-energy) band. */
export function qwzBerryCurvature(kx, ky, parameters = {}) {
  const mass = parameters.mass ?? -1;
  const hopping = parameters.hopping ?? 1;
  const trBreaking = parameters.trBreaking ?? 1;
  const vector = qwzD(kx, ky, parameters);
  const length = norm3(vector);
  if (length < EPSILON) return null;

  const numerator = hopping * trBreaking * (
    Math.sin(kx) ** 2 * Math.cos(ky)
    + Math.cos(kx) * Math.sin(ky) ** 2
    + (mass + Math.cos(kx) + Math.cos(ky)) * Math.cos(kx) * Math.cos(ky)
  );
  return numerator / (2 * length ** 3);
}

/** Exact phase classification away from a bulk gap closing. */
export function qwzChernNumber(parameters = {}) {
  const mass = parameters.mass ?? -1;
  const hopping = parameters.hopping ?? 1;
  const trBreaking = parameters.trBreaking ?? 1;
  for (const [value, name] of [
    [mass, "mass"],
    [hopping, "hopping"],
    [trBreaking, "trBreaking"],
  ]) assertFinite(value, name);
  if ([-2, 0, 2].some(critical => Math.abs(mass - critical) < EPSILON)) return null;
  const flattened = Math.abs(hopping) < EPSILON || Math.abs(trBreaking) < EPSILON;
  if (flattened && Math.abs(mass) < 2) return null;
  if (Math.abs(mass) > 2 || flattened) return 0;
  const orientation = Math.sign(hopping * trBreaking);
  return mass < 0 ? orientation : -orientation;
}

function orientedSolidAngle(a, b, c) {
  const numerator = dot3(a, cross3(b, c));
  const denominator = 1 + dot3(a, b) + dot3(b, c) + dot3(c, a);
  return 2 * Math.atan2(numerator, denominator);
}

/** Gauge-free lattice estimate from oriented spherical triangle areas. */
export function numericalQwzChern(parameters = {}, resolution = 45) {
  if (!Number.isInteger(resolution) || resolution < 4) {
    throw new RangeError("resolution must be an integer >= 4");
  }
  const points = Array.from({length: resolution}, (_, ix) => (
    Array.from({length: resolution}, (_, iy) => unitD(
      -Math.PI + TAU * ix / resolution,
      -Math.PI + TAU * iy / resolution,
      parameters,
    ))
  ));
  if (points.some(row => row.some(point => point === null))) return null;

  let solidAngle = 0;
  for (let ix = 0; ix < resolution; ix += 1) {
    for (let iy = 0; iy < resolution; iy += 1) {
      const a = points[ix][iy];
      const b = points[(ix + 1) % resolution][iy];
      const c = points[(ix + 1) % resolution][(iy + 1) % resolution];
      const d = points[ix][(iy + 1) % resolution];
      solidAngle += orientedSolidAngle(a, b, c) + orientedSolidAngle(a, c, d);
    }
  }
  return -solidAngle / (4 * Math.PI);
}

/** Grid estimate of the full gap, with all known closing manifolds exact. */
export function qwzBandGap(parameters = {}, resolution = 121) {
  const mass = parameters.mass ?? -1;
  const hopping = parameters.hopping ?? 1;
  const trBreaking = parameters.trBreaking ?? 1;
  if ([-2, 0, 2].some(critical => Math.abs(mass - critical) < EPSILON)) return 0;
  const flattened = Math.abs(hopping) < EPSILON || Math.abs(trBreaking) < EPSILON;
  if (flattened && Math.abs(mass) <= 2) return 0;

  let minimum = Infinity;
  for (let ix = 0; ix < resolution; ix += 1) {
    const kx = -Math.PI + TAU * ix / (resolution - 1);
    for (let iy = 0; iy < resolution; iy += 1) {
      const ky = -Math.PI + TAU * iy / (resolution - 1);
      minimum = Math.min(minimum, qwzEnergy(kx, ky, parameters));
    }
  }
  return 2 * minimum;
}

export function qwzBandPath(parameters = {}, samplesPerSegment = 80) {
  const points = [
    {label: "Γ", kx: 0, ky: 0},
    {label: "X", kx: Math.PI, ky: 0},
    {label: "M", kx: Math.PI, ky: Math.PI},
    {label: "Y", kx: 0, ky: Math.PI},
    {label: "Γ", kx: 0, ky: 0},
  ];
  const samples = [];
  for (let segment = 0; segment < points.length - 1; segment += 1) {
    const start = points[segment];
    const end = points[segment + 1];
    for (let index = 0; index < samplesPerSegment; index += 1) {
      const fraction = index / samplesPerSegment;
      const kx = start.kx + fraction * (end.kx - start.kx);
      const ky = start.ky + fraction * (end.ky - start.ky);
      samples.push({x: segment + fraction, energy: qwzEnergy(kx, ky, parameters)});
    }
  }
  samples.push({x: 4, energy: qwzEnergy(0, 0, parameters)});
  return {samples, ticks: points.map((point, index) => ({x: index, label: point.label}))};
}

export function sshD(k, parameters = {}) {
  const t1 = parameters.t1 ?? 0.65;
  const t2 = parameters.t2 ?? 1;
  for (const [value, name] of [[k, "k"], [t1, "t1"], [t2, "t2"]]) {
    assertFinite(value, name);
  }
  return {x: t1 + t2 * Math.cos(k), y: t2 * Math.sin(k), z: 0};
}

export function sshEnergy(k, parameters = {}) {
  const vector = sshD(k, parameters);
  return Math.hypot(vector.x, vector.y);
}

export function sshBandGap(parameters = {}) {
  const t1 = parameters.t1 ?? 0.65;
  const t2 = parameters.t2 ?? 1;
  return 2 * Math.abs(Math.abs(t1) - Math.abs(t2));
}

export function sshWindingNumber(parameters = {}) {
  const t1 = parameters.t1 ?? 0.65;
  const t2 = parameters.t2 ?? 1;
  if (Math.abs(Math.abs(t1) - Math.abs(t2)) < EPSILON) return null;
  return Math.abs(t1) < Math.abs(t2) ? 1 : 0;
}

export function sshPath(parameters = {}, samples = 241) {
  if (!Number.isInteger(samples) || samples < 3) throw new RangeError("samples must be >= 3");
  return Array.from({length: samples}, (_, index) => {
    const k = -Math.PI + TAU * index / (samples - 1);
    const d = sshD(k, parameters);
    return {k, ...d, energy: Math.hypot(d.x, d.y)};
  });
}

export function sshFiniteHamiltonian(cells = 18, parameters = {}) {
  if (!Number.isInteger(cells) || cells < 2) throw new RangeError("cells must be >= 2");
  const t1 = parameters.t1 ?? 0.65;
  const t2 = parameters.t2 ?? 1;
  const size = 2 * cells;
  const matrix = Array.from({length: size}, () => Array(size).fill(0));
  for (let site = 0; site < size - 1; site += 1) {
    const hopping = site % 2 === 0 ? t1 : t2;
    matrix[site][site + 1] = hopping;
    matrix[site + 1][site] = hopping;
  }
  return matrix;
}

/** Jacobi diagonalization for the small real-symmetric matrices used here. */
export function diagonalizeSymmetric(matrix, tolerance = 1e-12, maxSweeps = 80) {
  const size = matrix.length;
  if (!size || matrix.some(row => row.length !== size)) {
    throw new RangeError("matrix must be non-empty and square");
  }
  const a = matrix.map(row => row.slice());
  const vectors = Array.from({length: size}, (_, row) => (
    Array.from({length: size}, (_, column) => Number(row === column))
  ));

  for (let sweep = 0; sweep < maxSweeps; sweep += 1) {
    let maximum = 0;
    for (let p = 0; p < size - 1; p += 1) {
      for (let q = p + 1; q < size; q += 1) {
        const apq = a[p][q];
        maximum = Math.max(maximum, Math.abs(apq));
        if (Math.abs(apq) < tolerance) continue;
        const tau = (a[q][q] - a[p][p]) / (2 * apq);
        const tangent = tau === 0
          ? 1
          : Math.sign(tau) / (Math.abs(tau) + Math.sqrt(1 + tau * tau));
        const cosine = 1 / Math.sqrt(1 + tangent * tangent);
        const sine = tangent * cosine;
        const app = a[p][p];
        const aqq = a[q][q];

        for (let row = 0; row < size; row += 1) {
          if (row === p || row === q) continue;
          const arp = a[row][p];
          const arq = a[row][q];
          a[row][p] = cosine * arp - sine * arq;
          a[p][row] = a[row][p];
          a[row][q] = sine * arp + cosine * arq;
          a[q][row] = a[row][q];
        }
        a[p][p] = cosine ** 2 * app - 2 * sine * cosine * apq + sine ** 2 * aqq;
        a[q][q] = sine ** 2 * app + 2 * sine * cosine * apq + cosine ** 2 * aqq;
        a[p][q] = 0;
        a[q][p] = 0;

        for (let row = 0; row < size; row += 1) {
          const vrp = vectors[row][p];
          const vrq = vectors[row][q];
          vectors[row][p] = cosine * vrp - sine * vrq;
          vectors[row][q] = sine * vrp + cosine * vrq;
        }
      }
    }
    if (maximum < tolerance) break;
  }

  const order = Array.from({length: size}, (_, index) => index)
    .sort((left, right) => a[left][left] - a[right][right]);
  return {
    values: order.map(index => a[index][index]),
    vectors: order.map(column => vectors.map(row => row[column])),
  };
}

export function sshFiniteSpectrum(cells = 18, parameters = {}) {
  const decomposition = diagonalizeSymmetric(sshFiniteHamiltonian(cells, parameters));
  const centralIndices = Array.from({length: decomposition.values.length}, (_, index) => index)
    .sort((left, right) => (
      Math.abs(decomposition.values[left]) - Math.abs(decomposition.values[right])
    ))
    .slice(0, 2);
  const density = Array(2 * cells).fill(0);
  for (const index of centralIndices) {
    decomposition.vectors[index].forEach((amplitude, site) => {
      density[site] += amplitude ** 2 / centralIndices.length;
    });
  }
  const edgeWeight = density[0] + density[1] + density.at(-2) + density.at(-1);
  const inverseParticipation = density.reduce((sum, probability) => sum + probability ** 2, 0);
  return {
    ...decomposition,
    centralIndices,
    centralEnergies: centralIndices.map(index => decomposition.values[index]).sort((a, b) => a - b),
    density,
    edgeWeight,
    inverseParticipation,
  };
}

export {TAU};

File: visualizations/quantum-hall-band-topology/physics.py
"""Two-band Chern-insulator and SSH-model calculations.

The Qi--Wu--Zhang convention used throughout this visualization is

    d(k) = (A sin(kx), -lambda sin(ky), m + cos(kx) + cos(ky)).

The reported Chern number belongs to the negative-energy band.  With
``A * lambda > 0``, its sequence as ``m`` increases is ``0, +1, -1, 0``.
This sign uses ``A_i = -i <u|partial_i u>`` and hence
``Omega_- = -dhat dot (partial_x dhat cross partial_y dhat) / 2``.
The SSH chain starts and ends with the intracell ``t1`` bond.
"""

from __future__ import annotations

import math
from dataclasses import dataclass

import numpy as np

Array = np.ndarray
EPSILON = 1.0e-11


def _finite(*values: float) -> None:
    if not all(math.isfinite(value) for value in values):
        raise ValueError("model parameters must be finite")


def qwz_d(
    kx: Array | float,
    ky: Array | float,
    *,
    mass: float = -1.0,
    hopping: float = 1.0,
    tr_breaking: float = 1.0,
) -> Array:
    """Return the coefficient vector of the two-band QWZ Hamiltonian."""

    _finite(mass, hopping, tr_breaking)
    kx_array, ky_array = np.broadcast_arrays(
        np.asarray(kx, dtype=float), np.asarray(ky, dtype=float)
    )
    if np.any(~np.isfinite(kx_array)) or np.any(~np.isfinite(ky_array)):
        raise ValueError("momenta must be finite")
    return np.stack(
        (
            hopping * np.sin(kx_array),
            -tr_breaking * np.sin(ky_array),
            mass + np.cos(kx_array) + np.cos(ky_array),
        )
    )


def qwz_energy(
    kx: Array | float,
    ky: Array | float,
    *,
    mass: float = -1.0,
    hopping: float = 1.0,
    tr_breaking: float = 1.0,
) -> Array | float:
    """Return the positive energy ``|d|``; the two bands are ``+/-|d|``."""

    energy = np.linalg.norm(
        qwz_d(
            kx,
            ky,
            mass=mass,
            hopping=hopping,
            tr_breaking=tr_breaking,
        ),
        axis=0,
    )
    return float(energy) if energy.ndim == 0 else energy


def qwz_unit_d(
    kx: Array | float,
    ky: Array | float,
    *,
    mass: float = -1.0,
    hopping: float = 1.0,
    tr_breaking: float = 1.0,
) -> Array:
    """Return the normalized ``d`` vector, with ``nan`` at a band touching."""

    vector = qwz_d(
        kx,
        ky,
        mass=mass,
        hopping=hopping,
        tr_breaking=tr_breaking,
    )
    length = np.linalg.norm(vector, axis=0)
    return np.divide(
        vector,
        length,
        out=np.full_like(vector, np.nan, dtype=float),
        where=length > EPSILON,
    )


def qwz_berry_curvature(
    kx: Array | float,
    ky: Array | float,
    *,
    mass: float = -1.0,
    hopping: float = 1.0,
    tr_breaking: float = 1.0,
) -> Array | float:
    """Return occupied-band Berry curvature ``Omega_-(kx, ky)``."""

    kx_array, ky_array = np.broadcast_arrays(
        np.asarray(kx, dtype=float), np.asarray(ky, dtype=float)
    )
    vector = qwz_d(
        kx_array,
        ky_array,
        mass=mass,
        hopping=hopping,
        tr_breaking=tr_breaking,
    )
    derivative_x = np.stack(
        (
            hopping * np.cos(kx_array),
            np.zeros_like(kx_array),
            -np.sin(kx_array),
        )
    )
    derivative_y = np.stack(
        (
            np.zeros_like(ky_array),
            -tr_breaking * np.cos(ky_array),
            -np.sin(ky_array),
        )
    )
    cross_product = np.cross(derivative_x, derivative_y, axisa=0, axisb=0, axisc=0)
    numerator = -0.5 * np.sum(vector * cross_product, axis=0)
    length = np.linalg.norm(vector, axis=0)
    curvature = np.divide(
        numerator,
        length**3,
        out=np.full_like(length, np.nan, dtype=float),
        where=length > EPSILON,
    )
    return float(curvature) if curvature.ndim == 0 else curvature


def qwz_chern_number(
    *, mass: float = -1.0, hopping: float = 1.0, tr_breaking: float = 1.0
) -> int | None:
    """Return the exact occupied-band phase label, or ``None`` if gapless."""

    _finite(mass, hopping, tr_breaking)
    if any(abs(mass - critical) < EPSILON for critical in (-2.0, 0.0, 2.0)):
        return None
    if (abs(hopping) < EPSILON or abs(tr_breaking) < EPSILON) and abs(mass) < 2.0:
        return None
    if abs(mass) > 2.0 or abs(hopping) < EPSILON or abs(tr_breaking) < EPSILON:
        return 0
    orientation = 1 if hopping * tr_breaking > 0 else -1
    return orientation if mass < 0.0 else -orientation


def _solid_angle(a: Array, b: Array, c: Array) -> float:
    numerator = float(np.dot(a, np.cross(b, c)))
    denominator = 1.0 + float(np.dot(a, b) + np.dot(b, c) + np.dot(c, a))
    return 2.0 * math.atan2(numerator, denominator)


def numerical_qwz_chern(
    *,
    mass: float = -1.0,
    hopping: float = 1.0,
    tr_breaking: float = 1.0,
    resolution: int = 45,
) -> float | None:
    """Estimate the occupied Chern number by oriented spherical areas."""

    if not isinstance(resolution, int) or resolution < 4:
        raise ValueError("resolution must be an integer greater than or equal to 4")
    momenta = -math.pi + 2.0 * math.pi * np.arange(resolution) / resolution
    kx, ky = np.meshgrid(momenta, momenta, indexing="ij")
    points = qwz_unit_d(
        kx,
        ky,
        mass=mass,
        hopping=hopping,
        tr_breaking=tr_breaking,
    )
    if np.any(~np.isfinite(points)):
        return None

    solid_angle = 0.0
    for ix in range(resolution):
        for iy in range(resolution):
            a = points[:, ix, iy]
            b = points[:, (ix + 1) % resolution, iy]
            c = points[:, (ix + 1) % resolution, (iy + 1) % resolution]
            d = points[:, ix, (iy + 1) % resolution]
            solid_angle += _solid_angle(a, b, c) + _solid_angle(a, c, d)
    return -solid_angle / (4.0 * math.pi)


def qwz_band_gap(
    *,
    mass: float = -1.0,
    hopping: float = 1.0,
    tr_breaking: float = 1.0,
    resolution: int = 181,
) -> float:
    """Return a grid estimate of the full bulk gap, exact at known closings."""

    _finite(mass, hopping, tr_breaking)
    if resolution < 5:
        raise ValueError("resolution must be at least 5")
    if any(abs(mass - critical) < EPSILON for critical in (-2.0, 0.0, 2.0)):
        return 0.0
    if (abs(hopping) < EPSILON or abs(tr_breaking) < EPSILON) and abs(mass) <= 2.0:
        return 0.0
    momenta = np.linspace(-math.pi, math.pi, resolution)
    kx, ky = np.meshgrid(momenta, momenta, indexing="ij")
    return 2.0 * float(
        np.min(
            qwz_energy(
                kx,
                ky,
                mass=mass,
                hopping=hopping,
                tr_breaking=tr_breaking,
            )
        )
    )


def ssh_d(k: Array | float, *, t1: float = 0.65, t2: float = 1.0) -> Array:
    """Return the SSH Bloch vector ``(t1+t2 cos k, t2 sin k, 0)``."""

    _finite(t1, t2)
    momentum = np.asarray(k, dtype=float)
    if np.any(~np.isfinite(momentum)):
        raise ValueError("momentum must be finite")
    return np.stack(
        (
            t1 + t2 * np.cos(momentum),
            t2 * np.sin(momentum),
            np.zeros_like(momentum),
        )
    )


def ssh_energy(k: Array | float, *, t1: float = 0.65, t2: float = 1.0) -> Array | float:
    """Return the positive SSH energy; the bands are ``+/-|d|``."""

    energy = np.linalg.norm(ssh_d(k, t1=t1, t2=t2), axis=0)
    return float(energy) if energy.ndim == 0 else energy


def ssh_band_gap(*, t1: float = 0.65, t2: float = 1.0) -> float:
    """Return the exact SSH bulk gap."""

    _finite(t1, t2)
    return 2.0 * abs(abs(t1) - abs(t2))


def ssh_winding_number(*, t1: float = 0.65, t2: float = 1.0) -> int | None:
    """Return the SSH winding for the declared unit cell, or ``None`` at closure."""

    _finite(t1, t2)
    if abs(abs(t1) - abs(t2)) < EPSILON:
        return None
    return int(abs(t1) < abs(t2))


def ssh_finite_hamiltonian(cells: int = 18, *, t1: float = 0.65, t2: float = 1.0) -> Array:
    """Return an open SSH chain beginning and ending with an intracell bond."""

    if not isinstance(cells, int) or cells < 2:
        raise ValueError("cells must be an integer greater than or equal to 2")
    _finite(t1, t2)
    matrix = np.zeros((2 * cells, 2 * cells), dtype=float)
    hoppings = np.where(np.arange(2 * cells - 1) % 2 == 0, t1, t2)
    indices = np.arange(2 * cells - 1)
    matrix[indices, indices + 1] = hoppings
    matrix[indices + 1, indices] = hoppings
    return matrix


@dataclass(frozen=True, slots=True)
class SSHFiniteSpectrum:
    """Finite-chain eigensystem and diagnostics for the two central states."""

    energies: Array
    eigenvectors: Array
    central_indices: Array
    central_energies: Array
    density: Array
    edge_weight: float
    inverse_participation: float


def ssh_finite_spectrum(cells: int = 18, *, t1: float = 0.65, t2: float = 1.0) -> SSHFiniteSpectrum:
    """Diagonalize the finite chain and average its two most central densities."""

    energies, eigenvectors = np.linalg.eigh(ssh_finite_hamiltonian(cells, t1=t1, t2=t2))
    central_indices = np.argsort(np.abs(energies), kind="stable")[:2]
    density = np.mean(np.abs(eigenvectors[:, central_indices]) ** 2, axis=1)
    edge_weight = float(np.sum(density[[0, 1, -2, -1]]))
    return SSHFiniteSpectrum(
        energies=energies,
        eigenvectors=eigenvectors,
        central_indices=central_indices,
        central_energies=np.sort(energies[central_indices]),
        density=density,
        edge_weight=edge_weight,
        inverse_participation=float(np.sum(density**2)),
    )
